### PR TITLE
All Translog inner closes should happen after tragedy exception is set

### PR DESCRIPTION
**Pull request by andrershov**
We faced with the nasty race condition. See https://github.com/elastic/elasticsearch/issues/32526
`InternalEngine.failOnTragic` method has thrown `AssertionError`.
If you carefully look at if branches in this method, you will spot that its only possible, if either Lucene `IndexWriter `has closed from inside or `Translog`, has closed from inside, but tragedy exception is not set.
For now, let us concentrate on the `Translog` class.
We found out that there are two methods in Translog - namely `rollGeneration` and `trimOperations` that are closing Translog in case of Exception without tragedy exception being set.
This pull request fixes these 2 methods. To fix it, I've decided to pull `tragedyException` from `TranslogWriter` up-to `Translog` class, because in these 2 methods IndexWriter could be innocent, but still Translog needs to be closed.
Also to protect us in the future and make sure `close` method is never called from inside `Translog` special assertion examining stack trace is added. Since we're still targeting Java 8 for runtime - no StackWalker API is used in the implementation.

### CODE REVIEW

**Review by ywelsch — CHANGES_REQUESTED** [review]
Thanks for the PR. I've left some comments. I think we should also beef up the testing here to ensure that we indeed always have a tragic exception at hand in case a tragic event happened. A good starting point for this is `TranslogTests#testWithRandomException`, where we can sprinkle `assertNotNull(failableTLog.getTragicException());` over the place (this would have actually found the `rollGeneration` one), and call some more of the methods on failableTLog (e.g. `trimOperations`) so that we get good coverage.

**Comment by bleskes** [comment]
@ywelsch please also discuss how far this can be backported (I only see labels for 7.0.0)

**Comment by andrershov** [comment]
@ywelsch 

- I've wrapped AtomicReference with the tragic exception with TragedyExceptionHolder to make sure we do not lose, but instead suppress subsequent exceptions
- Made stack checking method static
- Added a test for stack checking method - the best thing we can do here is subclassing and check for subclasses in stack checking method
- Added non-null tragic exception assertions to tests that might fail during rollGeneration and trimOperations.

Could you please review my changes?

File: server/src/test/java/org/elasticsearch/index/translog/TranslogTests.java [hunk 1]
```diff
@@ -2931,6 +2934,46 @@ public void testCloseSnapshotTwice() throws Exception {
         }
     }
 
+    public void testTranslogCloseInvariant() throws IOException {
+        //close method should never be called directly from Translog (the only exception is closeOnTragicEvent)
```
[the remark this answers is not in the document]
Comment (ywelsch) [reply server/src/test/java/org/elasticsearch/index/translog/TranslogTests.java]: also move this comment outside of the method

[the remark this answers is not in the document]
Comment (ywelsch) [reply server/src/test/java/org/elasticsearch/index/translog/TranslogTests.java]: finally, add an `assumeTrue("test only works with assertions enabled", Assertions.ENABLED);` to the beginning of the method

File: server/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java [hunk 2]
```diff
@@ -129,28 +133,13 @@ public static TranslogWriter create(ShardId shardId, String translogUUID, long f
...
-        } else if (tragedy != ex) {
-            // it should be safe to call closeWithTragicEvents on multiple layers without
-            // worrying about self suppression.
-            tragedy.addSuppressed(ex);
-        }
-        try {
-            close();
-        } catch (final IOException | RuntimeException e) {
-            ex.addSuppressed(e);
+        if (tragedy.setTragicException(ex)) {
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java]: why change the semantics here to only call close when setting the tragedy the first time? Let's keep the existing semantics and make `setTragicException` return void

File: server/src/main/java/org/elasticsearch/index/translog/TragicExceptionHolder.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+public class TragicExceptionHolder {
+    private final AtomicReference<Exception> tragedy = new AtomicReference<>();
+
+    public boolean setTragicException(Exception ex){
+        assert ex != null;
+        if (tragedy.compareAndSet(null, ex)) {
+            tragedy.set(ex);
+            return true;
+        } else {
+            if (tragedy.get() != ex) { //to ensure there is no self-suppression
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/index/translog/TragicExceptionHolder.java]: add space after `//`

File: server/src/main/java/org/elasticsearch/index/translog/TragicExceptionHolder.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+import java.util.concurrent.atomic.AtomicReference;
+
+public class TragicExceptionHolder {
+    private final AtomicReference<Exception> tragedy = new AtomicReference<>();
+
+    public boolean setTragicException(Exception ex){
+        assert ex != null;
+        if (tragedy.compareAndSet(null, ex)) {
+            tragedy.set(ex);
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/index/translog/TragicExceptionHolder.java]: why double-set here? Isn't the previous line already doing this?

**Review by ywelsch — CHANGES_REQUESTED** [review]
Thanks for the extra test. I've left some more comments.

File: server/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java (hunk 2 above)
Comment (andrershov) [reply server/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java]: @ywelsch what is the reason to call close twice if every time we set tragic exception close is called?

**Comment by andrershov** [comment]
@ywelsch thank you for the review!
I've fixed everything except close once. I've commented on it.

File: server/src/main/java/org/elasticsearch/index/translog/TragicExceptionHolder.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+import java.util.concurrent.atomic.AtomicReference;
+
+public class TragicExceptionHolder {
+    private final AtomicReference<Exception> tragedy = new AtomicReference<>();
+
+    /**
+     * Sets the tragic exception or if the tragic exception is already set adds passed exception as suppressed exception
+     * @param ex tragic exception to set
+     */
+    public void setTragicException(Exception ex){
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/index/translog/TragicExceptionHolder.java]: space missing between `)` and `{`

File: server/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java [hunk 6]
```diff
@@ -129,31 +133,16 @@ public static TranslogWriter create(ShardId shardId, String translogUUID, long f
...
-        }
+        tragedy.setTragicException(ex);
         try {
             close();
         } catch (final IOException | RuntimeException e) {
             ex.addSuppressed(e);
         }
     }
 
+
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java]: remove extra newline

**Review by ywelsch — APPROVED** [review]
LGTM. Let's backport this to 6.4.1 and 6.5.0 as well.

**Comment by andrershov** [comment]
Merged to master, pending backport to 6.5.0 and 6.4.1

**Pull request merged**